### PR TITLE
chore: Remove flash-attention v1

### PR DESCRIPTION
**Pull request by ckadner**
#### Motivation

Building the wheel for the `flash-attention` and it's extension module dropout `layer_norm` takes more than 1 hour (`3702` sec). The `layer_norm` extension has been replaced in [v2.4.3](https://github.com/Dao-AILab/flash-attention/commit/abbc131). Now Flash Attention uses a [Triton-based implementation](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/ops/triton/layer_norm.py)

Flash-attention v1 was retained in TGIS in order to continue to [support older GPUs (NVIDIA T4, compute capability 7.5)](https://github.com/IBM/text-generation-inference/blob/c2eb7f7b3051a1616ad3d8db890288c38c18bb61/server/text_generation_server/utils/flash_attn.py#L9), which is no longer required.

Removing flash-attention v1 from the `server-release` image, the build could potentially be sped up by an hour.

#### Modifications

- [x] remove `flash-att-v1` build stage from Dockerfile
- [x] remove `server/Makefile-flash-att`
- [x] create GitHub package for [cache image on GitHub container registry](https://github.com/IBM/text-generation-inference/pkgs/container/text-gen-server)
- [x] push full cache image to ghcr,io on push to main (PR merged) 
- [x] use cache image from ghcr.io for PR builds
- [x] replace build stages/step with single `build-and-push` action
- [x] temporarily build Flash Attention extension module `dropout_layer_norm` and `rotary_emb` from flash-attention v2.5.2 sources (TODOs below)

**TODOs for future PR:**
- [ ] replace usages of `dropout_layer_norm` v1 with [Triton-based impl in v2](https://github.com/Dao-AILab/flash-attention/commit/abbc131)
- [ ] replace usages of `rotary_emb` v1 wit [Triton-based impl in v2](https://github.com/Dao-AILab/flash-attention/commit/942fcbf)
- [ ] remove `dropout_layer_norm` and `rotary_emb` module builds from Dockerfile

#### Result

Significantly shortened build times:
- Build/build time 95 min --> 25 min (full build) --> 2 min (cached build)
- Test/build time 6 min --> 2 min (full build) --> 45 sec (cached build)

So, overall, PR checks should complete in 25 min at most, but could be done in just over 6 min if most or all image layers can be reused from the cache image.

image

**Note:**

After (re)including `dropout_layer_norm` and `rotary_emb` module builds for Flash Attention -- albeit this time from v2 -- the full build took [75 min](https://github.com/IBM/text-generation-inference/actions/runs/8072952960/job/22055709881#step:8:309) (using cached layers from `python-builder`). 

Using that new cache image, building again after merging the changes from #39 only took [5 mins](https://github.com/IBM/text-generation-inference/actions/runs/8075019499/job/22061246539).

#### Related Issues

N/A

### CODE REVIEW

**Comment by ckadner** [comment]
@njhill -- I think this is ready for review (and merge :-) now

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -21,102 +23,75 @@ defaults:
...
   build:
     runs-on: ubuntu-latest
+    permissions:
+      packages: write
+      contents: read
     env:
-      BUILDKIT_INLINE_CACHE: 1
+      CACHE_IMAGE: "ghcr.io/ibm/text-gen-server:build-cache"
+      CACHE_REGISTRY: "ghcr.io"
+      SERVER_IMAGE: "ghcr.io/ibm/text-gen-server:latest"  # TODO: don't push final image as a package to ghcr.io
```
Comment (joerunde) [inline .github/workflows/build.yml:37]: hmm, why not push the final image?

It'd be great to push both `text-gen-server:latest` and `text-gen-server:${short commit hash}` on main builds

Comment (ckadner) [reply .github/workflows/build.yml:37]: I don't think we should use GH packages for images that get pulled externally to honor the quotas
https://docs.github.com/en/billing/managing-billing-for-github-packages/about-billing-for-github-packages

**Review by joerunde — APPROVED** [review]
Let's get this merged!

**Pull request merged**